pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,150 @@ package heal
 )
 
-// RegisterClientFunc - required to inform heal server about new client connection and assign it to the connection ID
-type RegisterClientFunc func(context.Context, *networkservice.Connection, networkservice.MonitorConnectionClient)
+type connectionInfo struct {
+	mut    sync.Mutex
+	conn   *networkservice.Connection
+	active bool
+}
 
 type healClient struct {
-	ctx context.Context
-	cc  networkservice.MonitorConnectionClient
+	ctx      context.Context
+	cc       networkservice.MonitorConnectionClient
+	initOnce sync.Once
+	initErr  error
+	conns    connectionInfoMap
 }
 
-// NewClient - creates a new networkservice.NetworkServiceClient chain element that inform healServer about new client connection
+// NewClient - creates a new networkservice.NetworkServiceClient chain element that monitors its connections' state
+//             and calls heal server in case connection breaks if heal server is present in the chain
+//             - ctx - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
+//             - cc  - MonitorConnectionClient that will be used to watch for connection confirmations and breakages.
 func NewClient(ctx context.Context, cc networkservice.MonitorConnectionClient) networkservice.NetworkServiceClient {
 	return &healClient{
 		ctx: ctx,
 		cc:  cc,
 	}
 }
 
+func (u *healClient) init(ctx context.Context, conn *networkservice.Connection) error {
+	monitorClient, err := u.cc.MonitorConnections(u.ctx, &networkservice.MonitorScopeSelector{
+		PathSegments: []*networkservice.PathSegment{{Name: conn.GetCurrentPathSegment().Name}, {Name: ""}},
+	})
+	if err != nil {
+		return errors.Wrap(err, "MonitorConnections failed")
+	}
+
+	healConnection := requestHealConnectionFunc(ctx)
+	if healConnection == nil {
+		healConnection = func(conn *networkservice.Connection) {}
+	}
+	restoreConnection := requestRestoreConnectionFunc(ctx)
+	if restoreConnection == nil {
+		restoreConnection = func(conn *networkservice.Connection) {}
+	}
+
+	go u.listenToConnectionChanges(healConnection, restoreConnection, monitorClient)
+
+	return nil
+}
+
 func (u *healClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
-	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	conn := request.GetConnection()
+	u.initOnce.Do(func() {
+		u.initErr = u.init(ctx, conn)
+	})
+	// if initialization failed, then we want for all subsequent calls to Request() on this object to also fail
+	if u.initErr != nil {
+		return nil, u.initErr
+	}
 
-	registerClient := registerClientFunc(ctx)
-	if err == nil && registerClient != nil {
-		registerClient(u.ctx, conn, u.cc)
+	connInfo, _ := u.conns.LoadOrStore(conn.GetId(), &connectionInfo{
+		conn: conn.Clone(),
+		mut:  sync.Mutex{},
+	})
+	u.replaceConnectionPath(conn, connInfo)
+
+	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return nil, err
```

Comment:
If it is a first request for the connection, we need to remove `connInfo` from `u.conns`.